### ISSUE TITLE
Allow more file types, like SVG

### ISSUE THREAD

**Issue by engeir** [HakonHarnes/img-clip.nvim#83]
**Is your feature request related to a problem? Please describe.**

Using the [telescope picker](https://github.com/HakonHarnes/img-clip.nvim?tab=readme-ov-file#telescopenvim), I was actually expecting to be able to choose any file and that the path would be placed in the appropriate template (pdf, md, or whatever), but at least file types like SVG. Instead, when I pick for example SVG, "Content is not an image." is printed. I could only find in the code that `png`, `jpg`, and `jpeg` are allowed ([here](https://github.com/HakonHarnes/img-clip.nvim/blob/28a32d811d69042f4fa5c3d5fa35571df2bc1623/lua/img-clip/util.lua#L132C2-L134C40)).

I also notice in the tests [here](https://github.com/HakonHarnes/img-clip.nvim/blob/28a32d811d69042f4fa5c3d5fa35571df2bc1623/tests/clipboard_spec.lua#L54C1-L54C60), that you check for the type `image/png`.

In short, I did not quite understand how the allowed/disallowed file type is done, but types like `image/...`, e.g `image/svg+xml` should probably just work.

**Describe the solution you'd like**
An easier way of specifying file types, e.g. in the setup function.

**Comment by HakonHarnes** [comment]
Heisann!

Thanks for reporting. This should be doable, but I unfortunately don't have time to implement it at the moment. I'll look into this when I have more time. PRs are welcome.

**Comment by idvorkin** [comment]
Add webp too :) 

Wasn't sure if we could, but asking in case - can we do something hacky with process_cmd with the existing plugin ? @HakonHarnes

**Comment by HakonHarnes** [comment]
I think we can solve this issue in two ways:

The first approach is to add a new configuration option called `formats` which takes a list of allowed formats (e.g. `formats = {'png', 'jpg', 'svg', ...)`. The plugin will then check against those formats when embedding file paths. This is the simplest approach. 

The problem with this approach is that different file formats require different markup. For example, in markdown images are embedded with `![${CURSOR}](${FILE_PATH}]` while files are embedded with `[${CURSOR}](${FILE_PATH})` (notice the absence of the exclamation mark). 

The second approach could then be to define multiple templates and formats. For example: 

```lua
filetypes = {
  markdown = {
    templates = {
      image = "![${CURSOR}](${FILE_PATH})",
      file = "[${CURSOR}](${FILE_PATH})"
    },
    formats = {
      image = {"png", "jpg", "jpeg", "gif", "webp"},
      file = {"pdf", "doc", "docx", "html", "zip"} 
    }
  }
}
```

Of course, this makes the plugin more capable, but also more complex and harder to configure for the end user. I also feel like this extends the functionality of the plugin far enough that it should be renamed to something else. Perhaps `embed.nvim` or similar, since it would then embed any file, not just images from the clipboard.

In either case I'll have to improve how file extensions are handled. It currently just defaults to `.png`. It's easy to figure out which file extension is correct for file paths, but not so much for URLs and clipboard content, unfortunately (these usually return `image/png`, even though e.g. a `webp` file is copied).

I'm leaning towards the second approach, but I'll have to think about this further. I don't want to introduce breaking changes and complex configuration options for users if I can avoid it. Let me know your thoughts.

**Comment by chappertron** [comment]
Hi, I've had a go at implementing this. 

The approach I went for was adding an extra config option called `extra_image_types` that the user can use to specify additional types that will be treated as images. 

As such, the config looks like this:

```lua
    filetypes = {

      tex = {
        --- File types other than jpg and png to treat as images
        extra_image_types = { 'pdf' }, ---@type string[]
      },
    },
```

I have not implemented the different template types as in your second suggested spec. 
This does, however, at least allow for files that use the same template for embedding for both pdfs and pngs to support these additional file types.
An example of this is LaTeX, which commonly uses .pdf as a figure type.

I could rename the `extra_image_types` to `formats` as in your proposed configuration, then later on it could be expanded to support the image and file keys. 

Let me know what you think and if you would like me to submit a PR.

**Comment by HakonHarnes** [comment]
Hi, chappertron!

Thanks for taking a look at this. I'm definitely interested in PRs for this issue.

I'd prefer to solve this with multiple templates right away to avoid introducing breaking changes later. Perhaps this configuration structure would suffice:

```lua
filetypes = {
  markdown = {
    formats = {
      images = {
        extensions = { "png", "jpg", "jpeg" },
        template = "![$CURSOR]($FILE_PATH)",
      },
      files = {
        extensions = { "pdf" },
        template = "[${CURSOR}](${FILE_PATH})",
      }
    }
  }
}
```

The formats could be called anything. `images` and `files` are just examples, but since they're named this allows the user to only override one format while keeping the others.

I'm also wondering if we should allow file globbing in the extensions (e.g. `*`, `?` etc.) for more flexibility. This would allow users to specify e.g. "every file regardless of extension should use this template"

If you have suggestions for other configuration structures that may be better, I'd love to hear them!

**Comment by chappertron** [comment]
Hi Håkon.

I will make the PR shortly.

An idea I had for supporting this would be to put the matching in the keys to templates.
Seeing as there is only likely to be a handful of image formats for each template, I can't imagine it would be too verbose.

It would then simplify the formats table, especially if globs are used.

``` lua
filetypes = {
    markdown = {
        formats = { 'png', 'jpg', 'pdf' },
        template = {
          ['pdf'] = '[$CURSOR]($FILE_PATH)',
          ['*'] = '![$CURSOR]($FILE_PATH)',
        },
      },
    }
```

This approach might cause some problems with implementation, especially if multiple formats map to a given template, like so: 

``` lua
filetypes = {
      markdown = {
        formats = { 'png', 'jpg', 'pdf' },
        template = {
          [{ 'pdf', 'svg' }] = '[$CURSOR]($FILE_PATH)',
          ['*'] = '![$CURSOR]($FILE_PATH)',
        },
      },
    }
```

The above might cause the complications looking up the template by image format.
Perhaps this could be solved by flattening the key and copying the matching template for easier look up.
Meaning the template table would be transformed into something like this internally, after the config is read:

``` lua
template = {
          ['pdf'] = '[$CURSOR]($FILE_PATH)',
          ['svg'] = '[$CURSOR]($FILE_PATH)',
          ['png'] = '![$CURSOR]($FILE_PATH)',
          ['jpg'] = '![$CURSOR]($FILE_PATH)',
        },
```

I'm not sure how common it is for tables to be used as keys in neovim configs, so maybe the second snippet isn't the best approach.

**Comment by chappertron** [comment]
Sorry for the chain of replies.

> The problem with this approach is that different file formats require different markup. For example, in markdown images are embedded with `![${CURSOR}](${FILE_PATH}]` while files are embedded with `[${CURSOR}](${FILE_PATH})` (notice the absence of the exclamation mark).

Thinking about it further, I don't think this is actually an issue for the markdown example.

 At least in GitHub Flavoured Markdown (GFM), the syntax `[${CURSOR}](${FILE_PATH})` will only create a link to the file, and not try and embed it. 
You need the leading `!`. 

The same is also true for Obsidian, to embed a PDF there you use either the 'wiki-link' syntax or normal link syntax with a leading ! (i.e. `![[${FILE_PATH}]]` or `[${CURSOR}](${FILE_PATH})`).

GitHub does support embedding SVGs ([as can be seen in their own example](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#images)) with this format.
The image formats supported by GitHub seem to be [PNG, JPG, GIF, PSD, and SVG](https://docs.github.com/en/repositories/working-with-files/using-files/working-with-non-code-files#rendering-and-diffing-images).
It does not, however, support embedding PDFs.

By contrast, Obsidian does support PDFs being embedded inline.

This seems to suggest to me that, at least for Markdown, the template format for embedding is the same for all image types. It is just down to the support of the implementation to render or embed that image/file.

That being said, GFM and Obsidian markdown are the only two I am really familiar with, so it could be SVGs or other formats are embedded entirely differently in other implementations. 

However, if you wanted to support different templates for other paste targets, then the multi template thing may be necessary.
For example, if you wanted to support dragging and dropping PDFs into HTML, that would certainly require extra templates. 
That does, however, start the process of the plugin evolving more into `embed.nvim`, as you have said.

Perhaps that could be a feature to add further along, though.
The config I suggested above would allow for a backwards compatible method of specifying templates and formats.
`template` could later become a `table|string` instead of just a `string`.

In the meantime, the `formats` table could be added and rely on the plugin users to know what formats are treated as images in their default embedding.
`formats` could be even made to have appropriate defaults for different destination files.

Sorry for the wall of text. I hope it got my point across clearly, though.

**Comment by HakonHarnes** [comment]
> An idea I had for supporting this would be to put the matching in the keys to templates.

I like this approach. I don't think we need the `formats` option since the file format / extension is specified in the keys of the template. If a user defines a template table, it should replace the entire default template table. Otherwise the user won't be able to "disable" the default templates provided.

> Perhaps this could be solved by flattening the key and copying the matching template for easier look up.

Sounds sensible. It will make lookups simpler, seeing as you can just do e.g. `template["pdf"]`. However, how do we implement globbing with this approach (e.g. `template["*"]` or `template["ps*"]`? We could of course just limit globbing to just accepting a single `*` as a default value and doing `template["pdf"] or template["*"]` or similar, albeit a bit limiting. 

> Thinking about it further, I don't think this is actually an issue for the markdown example.

You're probably right. It's still an issue for other formats, as you mention, like HTML. Another examples is LaTeX where I'd imagine some users would like to embed PDFs as figures:

```tex
\includegraphics{file.pdf}
```

While others want to include the PDF as pages: 

```tex
\includepdf[pages=-]{file.pdf}
```

> That does, however, start the process of the plugin evolving more into `embed.nvim`, as you have said.

I'd like to go down this route. "img-clip.nvim" is no longer descriptive of what this plugin does, even it current state. It's called img-clip because it was just pasting images from the clipboard, but it's a lot more powerful now.

**Comment by chappertron** [comment]
Hi @HakonHarnes 
> I don't think we need the formats option since the file format / extension is specified in the keys of the template. If a user defines a template table, it should replace the entire default template table.

I disagree. The `formats` option would be needed for supporting the old style of templates, that is a single template for every single format. It wouldn't be possible to specify formats separately then. This would make the case of lots of formats but only one method of embedding more concise.


I also think `formats` would be needed to support 'globbing', whilst also being able to limit the formats that can be used. 
In a way maybe it should not be truly a glob, but really a key special for a fallback template.
It can then be used for formats that are specified in `formats`, but don't have their own unique template.
I think literal globbing would be a problem because you don't want every file format to be accepted!

Perhaps though, both could be done. `formats` could be extended by the keys of `template` when it is a table, so a user doesn't need to specify both or get an error if they put a template for a format they didn't put in formats. 

By the way, for your LaTeX example, PDF is usually the default file format most people use for figures, so I don't think that `\includepdf` should be made the default template for LaTeX. I do agree the option would be nice, however.

**Comment by HakonHarnes** [comment]
Indeed, it does make sense to keep the formats option. It allows for backwards compatibility and simpler configurations in most cases. 

The formats list should be extended by the keys of the templates. Regarding globbing, the default fallback approach (i.e. only supporting *) will cover most cases. 

Thank you for taking the time to discuss this. You can proceed with the PR whenever you like.
